Adds device context manager

dc-dc-dc opened the pull request:
## Proposed changes

Utilize python contextlib to easily swap out what device is being used without having to keep track, updated fft tests to showcase effectiveness.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented:
Very nice! I've been wanting to do something like this on the C++ side to simplify writing ops which take streams

dc-dc-dc commented:
Yeah its pretty handy when you only have a device specific implementation. In c++ could probably achieve something similar with a lambda + macro

angeloskath commented:
Love it. We have been putting off doing that for a while now.

If you plan on doing the same in C++ then I 'd suggest RAII instead of lambda. Basically a simple struct that keeps the old device and sets the desired device on construction and resets to the old device on destruction. Managing device scopes is literally done with C++ scopes after that. Also this should be able to take a stream instead of simply a device so we can use it in the ops or gradients to simplify things if we want to.

awni commented:
💯 I was about to make the exact same comment re stream and RAII on the C++ side.

I don't want to overly complexify things, but I also wonder if we should do it on the C++ side and bind it in core instead of putting it in `mlx.utils` 🤔 . Seems sufficiently useful.

dc-dc-dc commented:
Something like this ? (kindof annoying the signatures dont match as default_device returns a const ref)
```c++
struct StreamContextManager {
  public:
  StreamContextManager(StreamOrDevice s) {
    _stream = default_stream(default_device());
    set_default_stream(to_stream(s));
  }

  ~StreamContextManager() {
    set_default_stream(_stream);
  }

  private:
    Stream _stream;
};
```

dc-dc-dc commented:
Pushed some progress, though it seems like after calling `set_default_device` in c++ it doesn't seem to have an affect from python. I can confirm that the stream is cpu in context manager but the subsequent ops calls still use gpu. Will mess with this more, but just pushing in case this is something you guys ran into before

dc-dc-dc commented:
nvm, fixing it now

dc-dc-dc commented:
Updated and added docs, works nicely with the `with` keyword. 

Couldn't get the decorator part working as I'd like, but it's not essential so will just leave it as is for now. If anyone is interested in adding decorator support, essentially just need to add `__call__` def to the `PyStreamContextManager` class that takes in a `py::function`  argument and return a wrapper function that calls `enter` `func(*args, **kwargs)` `exit` in that order.

angeloskath commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,63 @@
+        A context manager for setting the current device and stream.
+    
+        This class is a Python wrapper around the C++ class
+        `mlx::core::StreamContextManager`. It is used to set the current device
```
I don't think there is a need for that you can directly say that it can change the default device and stream for the duration of the context as you have it. No need to mention that it wraps the equivalent to the C++ object.

angeloskath commented on python/src/utils.cpp:
````diff
@@ -0,0 +1,63 @@
+    
+        Example:
+        ```python
+        import mlx.utils as mlx
````
```python
import mlx.core as mx
```

angeloskath commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,63 @@
+        import mlx.utils as mlx
+    
+        # Create a context manager for the current device and stream.
+        with mlx.StreamContextManager(mx.cpu):
```
Same typo as above.

angeloskath commented on python/mlx/utils.py:
```diff
@@ -1,5 +1,4 @@
 # Copyright © 2023 Apple Inc.
-
```
Probably removed unnecessarily when removing the original Python implementation of the context manager.

angeloskath reviewed (changes requested):
Love it! Left a couple of comments for some typos but otherwise looks great!

angeloskath commented on mlx/utils.h:
```diff
@@ -9,6 +9,27 @@
 
 namespace mlx::core {
 
+using StreamOrDevice = std::variant<std::monostate, Stream, Device>;
```
The header `<variant>` is needed for gcc to compile this. Also the same explicit header include should be removed from `ops.h` now unless it `std::variant` is used elsewhere.

(Didn't see the failed linux build when I did the initial review hence the separate comment)

awni commented on python/src/utils.cpp:
````diff
@@ -0,0 +1,59 @@
+  py::class_<PyStreamContextManager>(m, "StreamContextManager", R"pbdoc(
+        A context manager for setting the current device and stream.
+    
+        Example:
+        ```python
+        import mlx.core as mx
+    
+        # Create a context manager for the current device and stream.
+        with mx.StreamContextManager(mx.cpu):
+            # Run some code that uses the current device and stream.
+            pass
+        ```
````
Could you move this to below the `Args` section?

Also you need to use an RST code block here:

```
```suggestion
        Example:
        
        .. code-block:: python
            
            import mlx.core as mx
    
            # Create a context manager for the current device and stream.
            with mx.StreamContextManager(mx.cpu):
                # Run some code that uses the current device and stream.
                pass
```

awni commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,59 @@
+using namespace mlx::core;
+
+// Slightly different from the original, with python context on init we are not
+// in the context yet. Only create the inner context on enter then de   lete on
```
```suggestion
// in the context yet. Only create the inner context on enter then delete on
```

awni commented:
Yea right it's not so elegant on the implementation side. I don't love that, though I think it could certainly be done and without breaking any downstream code.

But let's at least consider alternative names. For example another option is lower case stream is the context manager:

```
with mx.stream(...):
   ....
```

Personally I like that a lot more than `mx.StreamConstextManager` even though it introduces a bit of ambiguity between `Stream` and `stream`.

Any other suggestions?

dc-dc-dc commented:
I think dropping the manager part and leaving it as `StreamContext` fits its purpose. 

```python
with mx.StreamContext(mx.cpu):
   ...
```

awni commented:
> I think dropping the manager part and leaving it as `StreamContext` fits its purpose.

I still like simple `mx.stream` more but that is certainly an improvement.

Just for some points of comparison:

- PyTorch uses `torch.cuda.device(..)` as the context manager
- Jax uses `jax.default_device(..)` as the context manager

awni commented:
> For streams torch uses a class called StreamContext they have a wrapper called stream that returns an instance of it

Lol so they do both of our suggestions. What's the point of having both, wouldn't one just always use `torch.stream`?

dc-dc-dc commented:
added the `stream` wrapper function so we support both options similar to torch

awni commented:
Wonderful, thanks!! That's really nice!

Two more comments, then I think we are good:

- Could you update the C++ to be `StreamContext` as well? Seems better to be consistent there.
- Update the tests and docs to use the `stream` wrapper. I think it would be good to have that as the preferred option.

awni commented on docs/src/python/devices_and_streams.rst:
```diff
@@ -9,6 +9,7 @@ Devices and Streams
   :toctree: _autosummary
 
    Device
+   StreamContext
```
Add `stream` to do

awni commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,71 @@
+      [](StreamOrDevice s) { return PyStreamContext(s); },
+      "s"_a,
+      R"pbdoc(
+        Wrap around the Context-manager StreamContext that selects a given stream.
```
```suggestion
        A wrapper around :obj:`StreamContext` to select the given stream.
```

awni commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,71 @@
+        Wrap around the Context-manager StreamContext that selects a given stream.
+
+        Args:
+            s: The stream or device to set as the current device and stream.
```
```suggestion
            s: The stream or device to set as the default device and stream.
```

awni commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,71 @@
+            s: The stream or device to set as the current device and stream.
+        
+        Returns:
+            A context manager for setting the current device and stream.
```
```suggestion
            A context manager for setting the default device and stream.
```

awni commented on python/src/utils.cpp:
```diff
@@ -0,0 +1,71 @@
+          import mlx.core as mx
+      
+          # Create a context manager for the current device and stream.
+          with mx.StreamContext(mx.cpu):
```
Use `mx.stream` here.

```suggestion
          with mx.stream(mx.cpu):
```

awni commented:
I pushed a few changes to the docs to fix some issues there.

Also I noticed that the following runs:

```
with mx.stream(None):
  pass
```

That should probably not run I think.

angeloskath reviewed (approved):
👍
